### PR TITLE
Feature: Pull metadata UIDs from instance

### PR DESCRIPTION
**Pull request by nshandra**
### :pushpin: Issue
Task: [Pull UIDs from instance](https://app.clickup.com/t/2ew2g7y)

### :memo: Implementation
Adds the ability to retrieve the IDs for metadata present in a Google spreadsheet from a DHIS2 instance.
The metadata present in the spreadsheet acts as a filter.

This script makes sure that two elements can not have the same name, which is possible in a DHIS2 instance.
Its possible to specify a path for the CSV files with the PULL_METADATA_CSV_PATH env var.

- Spreadsheet tabs modifications: 
programRuleActions: Column: programRuleActionType is now called name, a note is added with this info.

### CODE REVIEW

**Review by jordibc — CHANGES_REQUESTED** [review]
Thanks @nshandra !

Let me mention a few issues other than the other comments:

* The files are not prettified (yarn prettify). Having vscode with extension Prettier + setting "Format on save" enabled is an easy way to do it.
* The code in download-ids.ts and upload-server.ts has a few things in common (including `loadSheet()` and `makeSeed()`. I think those could move to a separate file (in `utils/` for example) that they both use.

File: README.md [hunk 1]
```diff
@@ -18,10 +18,13 @@ DHIS2_BASE_URL=...
 DHIS2_USERNAME=...
 DHIS2_PASSWORD=...
 UPDATE_CATEGORY_OPTION_COMBOS=false # true or false
+UPDATE_SERVER=false # true or false
+PULL_METADATA_CSV_PATH=...
 ```
 
 The first time, we have to run `yarn install` to install all the dependencies.
-After that, you can just run `yarn start` to generate and update the metadata.
+After that, you can run `yarn update-server` to generate and update the metadata
```
[the remark this answers is not in the document]
Comment (nshandra) [reply README.md:26]: Changed to build-metadata.ts

File: README.md [hunk 2]
```diff
@@ -36,10 +39,15 @@ The script will generate a json file containing all the
...
 
+If you need to generate only the json set the `UPDATE_SERVER` option to `false`.
+
 If you want to manually upload the generated json file to a dhis2 instance,
 use its `Import/Export` app, go to `Metadata import` and use `Merge` as
 the "import strategy".
 
+The `PULL_METADATA_CSV_PATH` variable will store the path where the CSV files will
+be written. If empty, the current working directory will be used.
+
```
[the remark this answers is not in the document]
Comment (nshandra) [reply README.md]: Updated for clarity.

File: package.json [hunk 3]
```diff
@@ -6,8 +6,9 @@
     "author": "EyeSeeTea team",
     "scripts": {
         "prettify": "prettier \"./**/*.{js,jsx,json,css,ts,tsx}\" --write",
-        "build": "ncc build src/scripts/main.ts -m && cp dist/index.js $npm_package_name.js",
-        "start": "ts-node -P tsconfig.json src/scripts/main.ts"
+        "build": "ncc build src/scripts/upload-server.ts -m && cp dist/index.js $npm_package_name.js",
```
[the remark this answers is not in the document]
Comment (nshandra) [reply package.json]: Build command removed from `package.json`.

File: src/utils/pullMetadata.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
+import { D2Api } from "@eyeseetea/d2-api/2.34";
+import _ from "lodash";
+import { MetadataItem } from "../domain/entities/MetadataItem";
+import { Sheet } from "../domain/entities/Sheet";
+import { createObjectCsvWriter } from "csv-writer";
```
[the remark this answers is not in the document]
Comment (nshandra) [reply src/utils/pullMetadata.ts:5]: Added `csv-writer` package to `package.json`.

File: src/scripts/download-ids.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import { D2Api } from "@eyeseetea/d2-api/2.34";
+import { config } from "dotenv-flow";
+import { google } from "googleapis";
+import _ from "lodash";
+import { MetadataItem } from "../domain/entities/MetadataItem";
+import { Sheet } from "../domain/entities/Sheet";
+import { getUid } from "../utils/uid";
+import { pullMetadata } from "../utils/pullMetadata";
+
+async function download_ids() {
```
[the remark this answers is not in the document]
Comment (nshandra) [reply src/scripts/download-ids.ts:10]: Fixed it. Prettified `download-ids.ts` and `pullMetadata.ts`.

**Comment by nshandra** [comment]
After looking into it i have removed `programRuleActions` from `allowedTypesArray` by now.
First, because it lacks a name-like field, but also because it can store a lot of different actions (potentially of the same kind) so we would need to make a comparison of multiple fields and get the programRule Id.

Not sure if this merits the extra work, will bring it up to Nacho tomorrow.

**Comment by nshandra** [comment]
Add programRuleActions back, and fixed the unused name field in programRuleActions, its now the programRuleActionType field.

**Pull request merged**